OpenRA.Mods.Common/MapUtils/MultiBrush.cs

```diff
@@ -0,0 +1,398 @@
+		{
+			var brushes = new List<MultiBrush>();
+			foreach (var node in my.Nodes)
+				if (node.Key.Split('@')[0] == "MultiBrush")
+					brushes.Add(new MultiBrush(map, node.Value));
+			return brushes;
+		}
+
+		/// <summary>
+		/// Clone the brush. Note that this does not deep clone any ActorPlans.
+		/// </summary>
+		public MultiBrush Clone()
+		{
+			return new MultiBrush(this);
+		}
+
+		void UpdateShape()
+		{
+			var xys = new HashSet<CVec>();
+
+			foreach (var (xy, _) in tiles)
+				xys.Add(xy);
+
+			foreach (var actorPlan in actorPlans)
+				foreach (var cpos in actorPlan.Footprint().Keys)
+					xys.Add(new CVec(cpos.X, cpos.Y));
+
+			shape = xys.OrderBy(xy => (xy.Y, xy.X)).ToArray();
+		}
+
+		/// <summary>
+		/// Add tiles from a template, optionally with a given offset. By
+		/// default, it will be auto-offset such that the first tile is
+		/// under (0, 0).
+		/// </summary>
+		public MultiBrush WithTemplate(ushort templateId, CVec? offset = null)
+		{
+			var tileset = Map.Rules.TerrainInfo as ITemplatedTerrainInfo;
+			var templateInfo = tileset.Templates[templateId];
+			if (templateInfo.PickAny)
+				throw new ArgumentException("PickAny not supported - create separate obstacles instead.");
+			for (var y = 0; y < templateInfo.Size.Y; y++)
+				for (var x = 0; x < templateInfo.Size.X; x++)
+				{
+					var i = y * templateInfo.Size.X + x;
+					if (templateInfo[i] != null)
+					{
+						if (offset == null)
+							offset = new CVec(-x, -y);
+						var tile = new TerrainTile(templateId, (byte)i);
+						tiles.Add((new CVec(x, y) + (CVec)offset, tile));
+					}
+				}
+
+			UpdateShape();
+			return this;
+		}
+
+		/// <summary>Add a single tile at (0, 0).</summary>
+		public MultiBrush WithTile(TerrainTile tile)
+		{
+			tiles.Add((new CVec(0, 0), tile));
+			UpdateShape();
+			return this;
+		}
+
+		/// <summary>Add an actor at (0, 0).</summary>
+		public MultiBrush WithActor(ActorPlan actor)
+		{
+			actorPlans.Add(actor);
+			UpdateShape();
+			return this;
+		}
+
+		/// <summary>
+		/// <para>For all spaces occupied by the brush, add the given tile.</para>
+		/// <para>This is useful for adding a backing tile for actors.</para>
+		/// </summary>
+		public MultiBrush WithBackingTile(TerrainTile tile)
+		{
+			if (Area == 0)
+				throw new InvalidOperationException("No area");
+			foreach (var xy in shape)
+				tiles.Add((xy, tile));
+
+			return this;
+		}
+
+		/// <summary>Update the weight.</summary>
+		public MultiBrush WithWeight(float weight)
+		{
+			Weight = weight;
+			return this;
+		}
+
+		/// <summary>
+		/// <para>Paint tiles onto the map and/or add actors to actorPlans at the given location.</para>
+		/// <para>contract specifies whether tiles or actors are allowed to be painted.</para>
+		/// <para>If nothing could be painted, throws ArgumentException.</para>
+		/// </summary>
+		public void Paint(List<ActorPlan> actorPlans, CPos paintAt, Replaceability contract)
```

Comment:
Add `Map` as an argument here and pass it into `PaintTiles` to remove the dependency on `Map` in the constructor?